### PR TITLE
add validationwebhook for KubeVirt.Spec.CustomizeComponents

### PR DESCRIPTION
**Pull request by ashleyschuett**
Signed-off-by: Ashley Schuett <aschuett@redhat.com>

**What this PR does / why we need it**:

This adds validation to KV resources CustomizeComponents object. It lets the user know before applying that empty strings for the `ResourceName` and `ResourceType` are not valid. It will also alert the user if the `patch` that they have provided is not valid json. 

This is a follow up to PR #4984 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Add validation to CustomizeComponents object on the KubeVirt resource 
```

### CODE REVIEW

File: pkg/util/webhooks/validating-webhooks/validating-webhook.go [hunk 1]
```diff
@@ -27,6 +29,32 @@ func (*AlwaysPassAdmitter) Admit(*v1beta1.AdmissionReview) *v1beta1.AdmissionRes
 	return NewPassingAdmissionResponse()
 }
 
+func ToAdmissionResponse(causes []v1.StatusCause) *v1beta1.AdmissionResponse {
```
Comment (EdDev) [inline pkg/util/webhooks/validating-webhooks/validating-webhook.go:32]: This one looks like a `NewAdmissionResponse`.

File: pkg/virt-operator/webhooks/kubevirt-update-admitter.go [hunk 2]
```diff
@@ -39,16 +39,17 @@ type KubeVirtUpdateAdmitter struct {
 	Client kubecli.KubevirtClient
 }
 
-// NewKubeVirtUpdateAdmitter creates a KubeVirtUpdateAdmitter
-func NewKubeVirtUpdateAdmitter(client kubecli.KubevirtClient) *KubeVirtUpdateAdmitter {
+// NewKubeVirtAdmitter creates a KubeVirtUpdateAdmitter
+func NewKubeVirtAdmitter(client kubecli.KubevirtClient) *KubeVirtUpdateAdmitter {
```
Comment (EdDev) [inline pkg/virt-operator/webhooks/kubevirt-update-admitter.go:43]: Why did you drop the "Update" part? It does create a new `KubeVirtUpdateAdmitter`. At least, if you change one, you should change the other, no?

File: pkg/virt-operator/resource/generate/components/webhooks.go [hunk 3]
```diff
@@ -96,6 +96,7 @@ func NewOpertorValidatingWebhookConfiguration(operatorNamespace string) *v1beta1
 				Rules: []v1beta1.RuleWithOperations{{
 					Operations: []v1beta1.OperationType{
 						v1beta1.Update,
+						v1beta1.Create,
```
Comment (EdDev) [inline pkg/virt-operator/resource/generate/components/webhooks.go:99]: How this one sneaked in? Can you explain how is this related to this change?

File: pkg/virt-operator/webhooks/kubevirt-update-admitter.go [hunk 4]
```diff
@@ -93,28 +97,56 @@ func (admitter *KubeVirtUpdateAdmitter) validateWorkloadPlacementUpdate() ([]met
 	return []metav1.StatusCause{}, nil
 }
 
-func getAdmissionReviewKubeVirt(ar *v1beta1.AdmissionReview) (new *v1.KubeVirt, old *v1.KubeVirt, err error) {
+func getAdmissionReviewKubeVirt(ar *v1beta1.AdmissionReview) (new *v1.KubeVirt, old *v1.KubeVirt, isUpdate bool, err error) {
```
Comment (EdDev) [inline pkg/virt-operator/webhooks/kubevirt-update-admitter.go:100]: This function uses a mixture of both names returned values and in the body explicitly returns others. For example, `isUpdate` is not used.

Please choose one or the other, but not a mix as it harder to read what is really going on.

Another problem is that it returns way too many arguments, which is usually a sign that it does too many things.
Per what I see, if `old` is `nil`, it means that nothing got updated and the new bool is redundant. Can it be used at the caller instead?

**Review by EdDev — CHANGES_REQUESTED** [review]
Could you please populate the commit message with details of what has been changed here and the reason for it?

File: pkg/virt-operator/resource/generate/components/webhooks.go (hunk 3 above)
Comment (ashleyschuett) [reply pkg/virt-operator/resource/generate/components/webhooks.go:99]: Before the KubeVirt CR only needed to be validated on update since we where only validating workload placement policy and on creation any workload placement policy is valid. But the validation for `componentsCustomization` could be incorrect on create or update and needs to be check for both events.

File: pkg/virt-operator/webhooks/kubevirt-update-admitter.go (hunk 2 above)
Comment (ashleyschuett) [reply pkg/virt-operator/webhooks/kubevirt-update-admitter.go:43]: Thought I did a replace all. I have updated `KubeVirtUpdateAdmitter` to be `KubevirtAdmitter`

File: pkg/virt-operator/resource/generate/components/webhooks.go (hunk 3 above)
Comment (davidvossel) [reply pkg/virt-operator/resource/generate/components/webhooks.go:99]: does "Create" help us here? The operator's validation webhook appears to be created after the KUbeVirt cr is created.

so, it seems the logic looks like. 
- post kubevirt cr
- install kubevirt (and operator validation webhook)

Validation of the kubevirt CR at creation time doesn't seem like it will ever happen due to us never actually having the webhook registered at the point of creation. 

does this sound accurate, or am i missing something

**Review by davidvossel — COMMENTED** [review]
I think coupling the registration of the kubevirt cr with the actual installtion process of kubevirt is creating a "chicken and egg" problem. we can't validate the kubevirt cr at creation time if the registration of the validation webhook occurs after the kubevirt cr is created.

File: pkg/virt-operator/resource/generate/components/webhooks.go (hunk 3 above)
Comment (ashleyschuett) [reply pkg/virt-operator/resource/generate/components/webhooks.go:99]: You are correct that the CR can be added and processed before the validation webhook is running... Hmm.. is there a better way to do this? My concern is the inconsistency that could happen in the following workflow. 

- User create KubeVirt CR with an invalid patch, it is never applied since it's invalid but the user doesn't know this.
- Five months later
- User goes to update KubeVirt CR setting to add featureflag or some other thing and now they get an error on update about a patch that has been there the whole time. 

Alternatively to this webhook validation I could add `// +kubebuilder:validation:MinLength=1` to the `ResourceType` and `ResourceName` fields to validate nonempty and then could actually get rid of this part of the validation hook completely. However valdiating the patch to be valid JSON i don't think can be done... we do log an error for this maybe that's enough. Do you think this inconsistency between update and create is okay? Or should just the logging be enough for this? or is there another solution I am not thinking of?

Comment (davidvossel) [reply pkg/virt-operator/resource/generate/components/webhooks.go:99]: > User create KubeVirt CR with an invalid patch, it is never applied since it's invalid but the user doesn't know this.

If a broken json patch at creation time prevents the kubevirt installation from occurring, then it can't be ignored.

> Do you think this inconsistency between update and create is okay?

as long as the malformed json patch prevents the install and is reported, then I think this inconsistency will be okay.

So, at creation time, I'd expect a malformed json patch to result in the KubeVirt CR failing to be applied, which will give us a nice human readable string in some condition indicating that a patch is malformed.

And, at update time (assuming the validation webhook and virt-api are online)  we'll catch these issues immediately and not let a malformed json patch get persisted on the KubeVirt CR. 

It's not great, but I'm not seeing a more practical solution right now

Comment (ashleyschuett) [reply pkg/virt-operator/resource/generate/components/webhooks.go:99]: yeah.. unfortunate but sounds reasonable to me. I have update the PR for reflect these changes.

File: pkg/virt-operator/webhooks/kubevirt-update-admitter.go [hunk 5]
```diff
@@ -57,8 +58,10 @@ func (admitter *KubeVirtUpdateAdmitter) Admit(ar *v1beta1.AdmissionReview) *v1be
 		return resp
 	}
 
+	results = validateCustomizeComponents(newKV.Spec.CustomizeComponents)
+
```
Comment (davidvossel) [inline pkg/virt-operator/webhooks/kubevirt-update-admitter.go:62]: did you by chance check what happens if you post an initial KubeVirt CR with invalid json for customize components?

I'm not 100% sure, but since we will miss the invalid json on the CREATE, but catch it on the UPDATE, it's possible the KubeVirt CR's status can't be written to either. I'm just not sure how all this is wired up.  If we can't write to the status due to a JSON field in customizeComponents being invalid, then we'd never be able to report the issue on the kubevirt CR. 

if you haven't, can you make sure the kubevirt status can still be updated if the customizeComponents is invalid?

Comment (ashleyschuett) [reply pkg/virt-operator/webhooks/kubevirt-update-admitter.go:62]: I doubled checked and If the the kubevirt cr is created with invalid syntax the system will not be deplyed and the status of the kubevirt CR is still able to be updated. 
```
reason: DeploymentFailed
status: "False"
```

The KubeVirt spec is unable to be updated with 

```
{"component":"virt-operator","kind":"","level":"error","msg":"Could not update the KubeVirt resource.","name":"kubevirt","namespace":"kubevirt","pos":"kubevirt.go:659","reason":"admission webhook \"kubevirt-update-validator.kubevirt.io\" denied the request: patch \"{\\\"name\\\": \\\"not-valid}\" is not valid JSON","timestamp":"2021-03-12T14:56:39.470732Z","uid":"05a8bc49-92f7-4a2f-8ea6-4c68f5d9960f"}
```

but the status is.

**Review by davidvossel — APPROVED** [review]
/lgtm
/approve
